src/dotnet/commands/dotnet-build3/MSBuildForwardingApp.cs

```diff
@@ -0,0 +1,59 @@
+﻿using Microsoft.DotNet.Cli.Utils;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Microsoft.DotNet.Cli
+{
+    public class MSBuildForwardingApp : ForwardingApp
+    {
+        private static readonly string s_msbuildExeName = "msbuild.exe";
+
+        public MSBuildForwardingApp(string[] argsToForward) :
+            base(GetMSBuildExePath(), 
+                argsToForward,
+                depsFile: GetDepsFile(),
+                runtimeConfig: GetRuntimeConfig(),
+                environment: GetEnvironment()) { }
+
+        private static Dictionary<string, string> GetEnvironment()
+        {
+            return new Dictionary<string, string>
+            {
+                { "MSBuildExtensionsPath", GetDotnetMSBuildTargetsPath() },
+                { "DotnetHostPath", GetHostPath() }
```

Comment:
The issue here is not only the host to execute msbuild, but the host msbuild should use to execute csc.exe. 

The first problem is like you described but we need additional data once we're inside msbuild to execute csc.exe successfully.